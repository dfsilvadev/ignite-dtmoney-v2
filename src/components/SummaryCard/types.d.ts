import { ReactNode } from "react";

export type SummaryCardProps = {
  title?: string;
  icon?: ReactNode;
  value?: string;
  variant?: "default" | "success" | "danger";
  isLoading?: boolean;
};

export type CardStyleProps = Pick<SummaryCardProps, "variant">;
