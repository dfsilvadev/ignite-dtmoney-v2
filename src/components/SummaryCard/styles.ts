import styled, { DefaultTheme, css } from "styled-components";

import { CardStyleProps } from "./summaryCard";

const cardVariant = {
  default: (theme: DefaultTheme) => css`
    background: ${theme.colors["gray-600"]};
  `,
  success: (theme: DefaultTheme) => css`
    background: ${theme.colors["green-700"]};
  `,
  danger: (theme: DefaultTheme) => css`
    background: ${theme.colors["red-500"]};
  `
};

export const Card = styled.div<CardStyleProps>`
  ${({ theme, variant }) => css`
    border-radius: ${theme.border.radius};
    padding: 1.5rem;

    ${!!variant && cardVariant[variant](theme)}
  `}
`;

export const Header = styled.header`
  ${({ theme }) => css`
    display: flex;
    align-items: center;
    justify-content: space-between;
    color: ${theme.colors["gray-300"]};

    svg {
      font-size: 32px;
    }
  `}
`;

export const Text = styled.strong`
  display: block;
  margin-top: 1rem;
  font-size: 2rem;
`;
