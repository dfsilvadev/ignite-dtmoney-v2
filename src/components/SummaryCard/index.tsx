import * as S from "./styles";

import { SummaryCardProps } from "./types";

const SummaryCard = ({
  title,
  icon,
  value = "R$ 0,00",
  variant = "default"
}: SummaryCardProps) => {
  return (
    <S.Card variant={variant} role="card" aria-label="Summary Card">
      <S.Header>
        <h4>{title}</h4>

        {!!icon && icon}
      </S.Header>

      <S.Text>{value}</S.Text>
    </S.Card>
  );
};

export default SummaryCard;
