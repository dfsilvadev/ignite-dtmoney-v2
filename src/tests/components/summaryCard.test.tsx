import { screen } from "@testing-library/react";
import { renderWithTheme } from "../../utils/tests/renderWithTheme";

import SummaryCard from "../../components/SummaryCard";
import { ArrowCircleUp } from "phosphor-react";

describe("SummaryCard", () => {
  it("should render a title when passed", () => {
    renderWithTheme(<SummaryCard title="Entradas" />);

    expect(
      screen.getByRole("heading", { name: /entradas/i })
    ).toBeInTheDocument();
  });

  it("should render an icon when passed", () => {
    renderWithTheme(
      <SummaryCard
        title="Entradas"
        icon={
          <ArrowCircleUp weight="bold" color="#00B37E" data-testid="icon" />
        }
      />
    );

    expect(screen.getByTestId("icon")).toBeInTheDocument();
    expect(screen.getByTestId("icon")).toHaveStyle({
      color: "rgb(0, 179, 126);",
      width: "1rem;",
      height: "1rem;"
    });
  });

  it("should render a value when passed", () => {
    renderWithTheme(
      <SummaryCard
        title="Entradas"
        value="R$ 1.259,00"
        icon={
          <ArrowCircleUp
            weight="bold"
            size={32}
            color="#00B37E"
            data-testid="icon"
          />
        }
      />
    );

    expect(screen.getByText("R$ 1.259,00")).toBeInTheDocument();
  });

  it("should render R$0.00 if you don't pass any value", () => {
    renderWithTheme(
      <SummaryCard
        title="Entradas"
        icon={
          <ArrowCircleUp
            weight="bold"
            size={32}
            color="#00B37E"
            data-testid="icon"
          />
        }
      />
    );

    expect(screen.getByText("R$ 0,00")).toBeInTheDocument();
  });

  it("should render correctly", () => {
    const { container } = renderWithTheme(
      <SummaryCard
        title="Entradas"
        value="R$ 1.259,00"
        icon={
          <ArrowCircleUp
            weight="bold"
            size={32}
            color="#00B37E"
            data-testid="icon"
          />
        }
      />
    );

    expect(container).toMatchSnapshot();
  });
});
