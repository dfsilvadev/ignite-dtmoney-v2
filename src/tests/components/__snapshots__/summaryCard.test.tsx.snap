// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SummaryCard should render correctly 1`] = `
.c0 {
  border-radius: 0.25rem;
  padding: 1.5rem;
  background: #323238;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  color: #C4C4CC;
}

.c1 svg {
  font-size: 32px;
}

.c2 {
  display: block;
  margin-top: 1rem;
  font-size: 2rem;
}

<div>
  <div
    aria-label="Summary Card"
    class="c0"
    role="card"
  >
    <header
      class="c1"
    >
      <h4>
        Entradas
      </h4>
      <svg
        data-testid="icon"
        fill="#00B37E"
        height="32"
        viewBox="0 0 256 256"
        width="32"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="none"
          height="256"
          width="256"
        />
        <circle
          cx="128"
          cy="128"
          fill="none"
          r="96"
          stroke="#00B37E"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="24"
        />
        <polyline
          fill="none"
          points="94.1 121.9 128 88 161.9 121.9"
          stroke="#00B37E"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="24"
        />
        <line
          fill="none"
          stroke="#00B37E"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="24"
          x1="128"
          x2="128"
          y1="168"
          y2="88"
        />
      </svg>
    </header>
    <strong
      class="c2"
    >
      R$ 1.259,00
    </strong>
  </div>
</div>
`;
